### PR TITLE
[MXNET-81] Fix crash with mx.nd.ones

### PR DESCRIPTION
**Pull request by anirudh2290**
## Description ##

Reported by @haojin2

Execution of code like the following on device with only 8 gpus will crash the process instead of throwing an exception.
```
mx.nd.ones((2,2), ctx=mx.gpu(11))
```
The exception is thrown here: https://github.com/apache/incubator-mxnet/blob/master/src/engine/threaded_engine_perdevice.cc#L239 and since its not inside `ExecuteOprBlock`, there is no corresponding catch and terminate is called. 

Call cudaSetDevice and throw an exception before pushing to engine to avoid crash.

### CODE REVIEW

**Comment by CodingCat** [comment]
Hi, the community has passed to vote about associating the code changes with JIRA (https://lists.apache.org/thread.html/ab22cf0e35f1bce2c3bf3bec2bc5b85a9583a3fe7fd56ba1bbade55f@%3Cdev.mxnet.apache.org%3E)

We have updated the guidelines for contributors in https://cwiki.apache.org/confluence/display/MXNET/Development+Process, please ensure that you have created a JIRA at https://issues.apache.org/jira/projects/MXNET/issues/ to describe your work in this pull request and include the JIRA title in your PR as [MXNET-xxxx] your title where MXNET-xxxx is the jira id

File: src/imperative/imperative_utils.h [hunk 1]
```diff
@@ -70,12 +70,14 @@ inline Context GetContext(const nnvm::NodeAttrs& attrs,
...
     ctx = Context::Create(ctx.dev_mask(), ctx.dev_id);
   }
-#if !MXNET_USE_CUDA
   if (ctx.dev_mask() == gpu::kDevMask) {
+#if !MXNET_USE_CUDA
     LOG(INFO) << "GPU support is disabled. Compile MXNet with "
               << "USE_CUDA=1 to enable GPU support.";
-  }
+#else
+    mshadow::SetDevice<gpu>(ctx.dev_id);
```
Comment (piiswrong) [inline src/imperative/imperative_utils.h]: It makes no sense to call setdevice for the main thread. This adds overhead.

Comment (anirudh2290) [reply src/imperative/imperative_utils.h]: The call is intended to check if the device_id is available. The overhead should be pretty low as it is only setting the dev id corresponding to the thread and there is no synchronization taking place. The other option is to get the cuda device count and check if it the device falls in the range but that would have a similar overhead.

Comment (piiswrong) [reply src/imperative/imperative_utils.h]: Why can't this be handled as an async error in the engine? We can raise the error when user calls asnumpy

Comment (anirudh2290) [reply src/imperative/imperative_utils.h]: @piiswrong Where would the async error be raised first ? Currently only exception thrown inside ExecuteOprBlock(during operator exec) is handled. One of the arguments to ExecuteOprBlock is the runtime context and so the stream needs to be set based on device availability before calling ExecuteOprBlock. Therefore, I cannot add this when pushing to engine/handle it as async error in engine.

Comment (anirudh2290) [reply src/imperative/imperative_utils.h]: @piiswrong what do you suggest ?

Comment (piiswrong) [reply src/imperative/imperative_utils.h]: you can check it in PushAsync/Sync.

Comment (anirudh2290) [reply src/imperative/imperative_utils.h]: @piiswrong Done.

File: src/engine/threaded_engine.cc [hunk 2]
```diff
@@ -309,6 +309,13 @@ void ThreadedEngine::PushAsync(AsyncFn fn, Context exec_ctx,
                                int priority,
                                const char* opr_name,
                                bool wait) {
+#if MXNET_USE_CUDA
+  if (exec_ctx.dev_mask() == gpu::kDevMask) {
+    int device_count = 0;
+    cudaGetDeviceCount(&device_count);
```
Comment (piiswrong) [inline src/engine/threaded_engine.cc]: This shouldn't be done every time as it would introduce overhead

File: src/engine/threaded_engine.cc [hunk 3]
```diff
@@ -309,6 +309,13 @@ void ThreadedEngine::PushAsync(AsyncFn fn, Context exec_ctx,
...
+    CHECK_LT(exec_ctx.dev_id, device_count) << "Incorrect GPU device id";
```
Comment (piiswrong) [inline src/engine/threaded_engine.cc]: error message should be more descriptive. i.e. what the id is and why its illegal

File: src/engine/threaded_engine.h [hunk 4]
```diff
@@ -538,6 +538,11 @@ class ThreadedEngine : public Engine {
   std::shared_ptr<common::ObjectPool<VersionedVarBlock> > objpool_varblk_ref_;
   std::shared_ptr<common::ObjectPool<ThreadedVar> >       objpool_var_ref_;
 
+#if MXNET_USE_CUDA
+  /*! \brief Number of GPU devices available */
+  int device_count = 0;
```
Comment (piiswrong) [inline src/engine/threaded_engine.h]: `atomic<int> device_count_ = -1`

File: src/engine/threaded_engine.cc [hunk 5]
```diff
@@ -309,6 +309,18 @@ void ThreadedEngine::PushAsync(AsyncFn fn, Context exec_ctx,
                                int priority,
                                const char* opr_name,
                                bool wait) {
+#if MXNET_USE_CUDA
+  if (exec_ctx.dev_mask() == gpu::kDevMask) {
+    if (device_count <= 0) {
```
Comment (piiswrong) [inline src/engine/threaded_engine.cc]: device_count_ < 0

File: src/engine/threaded_engine.cc [hunk 6]
```diff
@@ -309,6 +309,18 @@ void ThreadedEngine::PushAsync(AsyncFn fn, Context exec_ctx,
...
+      cudaGetDeviceCount(&device_count);
```
Comment (piiswrong) [inline src/engine/threaded_engine.cc]: int tmp;
cudaGetDeviceCount(&tmp);
device_count_ = tmp;

**Comment by anirudh2290** [comment]
@piiswrong Thanks for your review! Is this good to merge ?

**Pull request merged**